File: Python3/프로그래머스/2/92342. 양궁대회/양궁대회.py
def solution(n, info):
    max_score = 0
    result = []
    
    # 어피치랑 라이언 점수 계산
    def calculate_score(ans):
        appech_score = 0
        lion_score = 0
        for idx, score in enumerate(info):
            if score + ans[idx] == 0:
                continue
            if score >= ans[idx]:
                # 어피치 승리
                appech_score += (10 - idx)
            else:
                # 라이언 승리
                lion_score += (10 - idx)
        return lion_score - appech_score
    
    def backtracking(ans, cnt, i):
        
        # 마지막 화살까지 쏘면 종료
        if i == n + 1: 
            # print(ans, cnt)
            if cnt == n:
                score = calculate_score(ans)
                
                nonlocal result
                nonlocal max_score
                
                if score > 0:
                    if max_score < score:
                        result = [ans[:]]
                        max_score = score
                    elif max_score == score:
                        result.append(ans[:])
            return

        ans[i] = min(n-cnt, info[i] + 1)
        backtracking(ans[:], cnt + ans[i], i+1)
        
        ans[i] = 0
        backtracking(ans[:], cnt + ans[i], i+1)
               
    backtracking([0] * 11, 0, 0)
    
    if not result:
        return [-1]
    if len(result) == 1:
        return result[0]
    
    result.sort(key= lambda x: (x[10],x[9],x[8],x[7],x[6],x[5],x[4],x[3],x[2],x[1],x[0]))
    return result[-1]

File: Swift/프로그래머스/2/92342. 양궁대회/양궁대회.swift
import Foundation

func solution(_ n:Int, _ info:[Int]) -> [Int] {
    var result = [[Int]]()
    var max_num = 0
    
    func backtracking(_ i: Int, _ lionCnt: Int, _ apeachScore: Int, _ lionScore: Int, _ ans: [Int]) {
//        print(i, lionCnt, apeachScore, lionScore)
        if i < 0 {
            let diff = lionScore - apeachScore
            if diff > 0 {
                if diff > max_num {
                    max_num = diff
                    result = [ans]
                } else if diff == max_num {
                    result.append(ans)
                }
            }
            return
        }
        
        // 라이언이 이기는 경우 (이기려면 어피치보다 하나 더 많이 쏘자)
        let lion = i == 0 ? n - lionCnt : info[10 - i] + 1
        if lionCnt + lion <= n {
            var temp = ans
            temp.append(lion)
            backtracking(i-1, lionCnt + lion, apeachScore, lionScore + i, temp)
            temp.popLast()
        }
        
        // 어피치가 이기는 경우 (라이언은 0만 쏴야됨 -> 화살 낭비 금지)
        let apeach = info[10-i] != 0 ? apeachScore + i : apeachScore
        var temp = ans
        temp.append(0)
        backtracking(i-1, lionCnt, apeach, lionScore, temp)
        temp.popLast()
    }
    
    backtracking(10, 0, 0, 0, [])
    
    // 결과가 여러개인 경우 정렬해주기
    for i in 0...10 {
        result.sort{$0[i] > $1[i]}
    }
    
    if result.isEmpty {
        return [-1]
    } else {
        return result[0]
    }
}